### DISCUSSION TITLE
How to go about scaling the graphics on the canvas equally?

### DISCUSSION THREAD

**Discussion by Lydanne** [fltk-rs/fltk-rs#1264] [Q&A]
I want to zoom the graph now, but didn't find an efficient way.
![image](https://user-images.githubusercontent.com/39021696/182063826-36f991f7-3c7a-4394-a78e-bceaa983ae7e.png)

**Comment by MoAlyousef** [comment]
How are you zooming? To zoom you need to resize the widget, also if you’re using an Offscreen, you need to call Offscreen::rescale in the widget’s draw method or resize_callback method.

**Reply by Lydanne** [reply]
No, I want to scale when drawing(frame.draw), I see there is a draw::scale method, but it has no effect

**Reply by MoAlyousef** [reply]
How are you calling scale?
It should be called before draw operations not after, like here for example (C++):
https://github.com/fltk/fltk/blob/84cf24948245d893cde9b3ef8c9eee614191de80/test/doublebuffer.cxx#L41

**Reply by Lydanne** [reply]
I have tried, not very good, I go back to write a demo it

**Reply by Lydanne** [reply]
> How are you calling scale? It should be called before draw operations not after, like here for example (C++): https://github.com/fltk/fltk/blob/84cf24948245d893cde9b3ef8c9eee614191de80/test/doublebuffer.cxx#L41

https://github.com/WumaCoder/fltk-demo/blob/8cce88770f48020b01a082effe3d5db537e669d5/examples/demo9.rs#L121
Here is my example

**Reply by MoAlyousef** [reply]
So fltk has fast drawing functions, these take integers and slow ones for complex shapes and these take floats. 
fl_scale works for complex shapes but is ignored in fast functions.
A simple example:
```rust
use fltk::{enums::*, prelude::*, *};

fn main() {
    let app = app::App::default();
    let mut wind = window::Window::default().with_size(800, 600);
    let mut frame1 = frame::Frame::new(0, 0, 400, 600, None);
    let mut frame2 = frame::Frame::new(400, 0, 400, 600, None);
    wind.end();
    wind.show();

    frame1.draw(|f| {
        let x = f.x();
        let y = f.y();
        let w = 100;
        let h = 100;
        let scale = 2.0;
        draw::set_draw_color(Color::Black);
        draw::push_matrix();
        draw::scale(scale); // ignored due to draw_rect being a "fast drawing function"
        draw::draw_rect(x, y, x + w, y + h);
        draw::pop_matrix();
    });

    frame2.draw(|f| {
        let x = f.x() as f64;
        let y = f.y() as f64;
        let w = 100.;
        let h = 100.;
        let scale = 2.0;
        draw::set_draw_color(Color::Black);
        draw::push_matrix();
        draw::translate(x - x * scale, y - y * scale);
        draw::scale(scale);
        draw::begin_line();
        draw::vertex(x, y);
        draw::vertex(x + w, y);
        draw::vertex(x + w, y + h);
        draw::vertex(x, y + h);
        draw::vertex(x, y);
        draw::end_line();
        draw::pop_matrix();
    });

    app.run().unwrap();
}
```

**Reply by Lydanne** [reply]
Thanks!
